Add reprepro-updater / reprepro for hosting apt repositories.

nuclearsandwich opened the pull request:
This change includes everything to get reprepro-updater and the apt repositories deployed for a build farm.

This is, by intention, pretty much a straight update of the [ros::repo module](https://github.com/ros-infrastructure/buildfarm_deployment/blob/master/modules/profile/manifests/ros/repo.pp) from buildfarm_deployment. There are some intentional changes:

* nginx is used for http rather than Apache. This is not made because Apache httpd was insufficient but because nginx is what the current maintainer (myself) is familiar with operationally.

* Repository suite and architecture information has been re-organized.
In the buildfarm_deployment_configuration, yaml-based hiera configs made embedding and maintaining file contents directly in hiera a viable option and the one that was used. However, it's not clear to those who aren't comfortable with the ROS build farm which items in the config file can be changed by end users and which are intended to be "static". Both the reprepro-updater ini config file and the default import_upstream configuration file have been moved to templates and the specific suites, architectures, and component are now chef attributes (which are roughly analogous to scalar puppet module parameters overridden via hiera). Default attributes are set in the cookbook but they can be superceded by attributes in the deployment environment.

I was partway through the implementation of pulp when I came up against deadlines that caused me to pause that work. There is a commit on this PR which has been applied and reverted and you'll notice the redis configuration file which has no purpose in the current branch. My intent is to resume work on getting the pulp work incorporated into chef after completing the initial migration. Unless something about the current state of the pulp work appears to interfere with the other running components I think  it can be ignored during this review as there will be a chance to review it in the subsequent revisions.

nuclearsandwich committed:
Address cookstyle feedback.

Mostly this is complaints about quote characters.

nuclearsandwich committed:
Execute bit is required for files to be created within the directory.

Although the mode in the buildfarm_deployment is 0644 the mode on the
live server is 0755.

nuclearsandwich committed:
Add repository management with reprepro and reprepro-updater.

This is ported somewhat directly from the existing puppet. The most
explicit change is the use of python3 rather than python2 for
initializing the repositories with reprepro-updater but since the
buildfarm scripts are all using python3 there is little likelihood for
issue.

nuclearsandwich committed:
Add a redis configuration based on Ubuntu's default.

nuclearsandwich committed:
Disable TCP and enable wide access via unix domain socket.

This allows processes on the machine to connect via Unix domain socket
with wide permissions. The socket can then be mounted to the containers
running pulp.

nuclearsandwich committed:
WIP on pulp support for repo recipe.

nuclearsandwich committed:
Revert "WIP on pulp support for repo recipe."

This reverts commit 65835e4fbad68aefc0763dce912e0fb5affb9a8a.

Taking this work out until we're ready for it.

nuclearsandwich committed:
Add nginx to serve repository files.

j-rivero commented on files/upload_repo.bash:
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+
```
```suggestion
set -e
```
stop the script on presence of errors

nuclearsandwich replied:
I'd actually prefer not to make this change now. The script has specific exit code handling for the ssh command from upstream. I can move it into a separate PR though.

j-rivero replied:
The `-e` should prevent us for possible failures in the script and return a success exit code. We can move to a different PR, it can affect heavily the behavior if some minor problems exist.

nuclearsandwich replied:
I fully agree that errexit / -e is better for any bash script. My concern is that if we *rely* on the script working during a failure right now and don't know it I don't want to change that during the migration window. Hence moving this out to anther PR.

nuclearsandwich replied:
Branch is up to be rebased for a future PR. https://github.com/ros-infrastructure/cookbook-ros-buildfarm/tree/upload-script-errexit

j-rivero commented on recipes/repo.rb:
```diff
@@ -1,49 +1,186 @@
+# .gnupg/gpg.conf
+directory "/home/#{agent_username}/.gnupg" do
+  owner agent_username
+  group agent_username
```
```suggestion
  group agent_username
  mode '0700'
```

nuclearsandwich replied:
Do you think it's important that the GPG *config* is not readable by other users?

nuclearsandwich replied:
Same as the .gpg directory I think we'd want this to be `0700` no?

j-rivero replied:
`0700` is executable bit no? In directories is being used to access items inside even if you can not list items inside, I think. Do we want that instead of complete blocking access to non user?

nuclearsandwich replied:
Executable bit on directories is required for the directory to be traversable. That is, in order to access foo/bar/baz both foo and bar need to be traversable.

j-rivero commented on recipes/repo.rb:
```diff
@@ -1,49 +1,186 @@
+cookbook_file "/home/#{agent_username}/.gnupg/gpg.conf" do
+  source 'gpg.conf'
+  owner agent_username
+  group agent_username
```
```suggestion
  group agent_username
  mode '0600'
```

nuclearsandwich replied:
Do you think it's important that the gpg config be unreadable by other system users?

j-rivero replied:
With the security kiddie hat on, I think the opposite: do we need to open it more than 0600? A malicious attack with a new/other user could potentially manipulate files in that directory.

j-rivero commented on recipes/repo.rb:
```diff
@@ -1,49 +1,186 @@
+gpg_key = data_bag_item('ros_buildfarm_repository_signing_keys', node.chef_environment)
+directory "/home/#{agent_username}/.ssh" do
+  owner agent_username
+  group agent_username
```
```suggestion
  group agent_username
  mode '0700'
```

nuclearsandwich replied:
Same as the .gpg directory I think we'd want this to be `0700` no?

j-rivero commented on recipes/repo.rb:
```diff
@@ -1,49 +1,186 @@
+    architectures: node['ros_buildfarm']['apt_repos']['architectures'],
+    signing_key: gpg_key['fingerprint'],
+    suites: node['ros_buildfarm']['apt_repos']['suites'],
+  ]
```
there is an `upstream_config` variable being used in the template that it is not explicit here. Might be a good idea for completeness to declare it and leave it blank if it is optional.

nuclearsandwich replied:
Thanks fixed in 3ade7d4

j-rivero commented on recipes/repo.rb:
```diff
@@ -1,49 +1,186 @@
+  ]
+end
+
+git "/home/#{agent_username}/reprepro-updater" do
```
I assume that although the repository does not have too much activity, we want chef to update to latest changes on every provisioning run.

nuclearsandwich replied:
The default behavior of this resource is to update.
From https://docs.chef.io/resources/git/

> :sync
>     Default. Update the source to the specified version, or get a new clone or checkout. This >  action causes a hard reset of the index and working tree, discarding any uncommitted changes.

j-rivero replied:
yes, sorry, I failed miserably in making my sentence a question: do we want to update on every chef run, don't we? (no changes required in that case, yes).

nuclearsandwich replied:
Yeah the intent is to always pull the latest. Good to check. In some cases I will specify the  the default action when I want to be explicit about what I want. Do you think this is a scenario where that makes sense?

j-rivero replied:
totally.

nuclearsandwich replied:
efd38a5

j-rivero commented on recipes/repo.rb:
```diff
@@ -1,49 +1,186 @@
+    environment 'PATH' => '/bin:/usr/bin', 'PYTHONPATH' => "/home/#{agent_username}/reprepro-updater/src", 'HOME' => "/home/#{agent_username}"
+    user agent_username
+    group agent_username
+    not_if "/usr/bin/python3 /home/#{agent_username}/reprepro-updater/scripts/setup_repo.py ubunut_#{repo} -q"
```
```suggestion
    not_if "/usr/bin/python3 /home/#{agent_username}/reprepro-updater/scripts/setup_repo.py ubuntu_#{repo} -q"
```

nuclearsandwich replied:
Well spotted.

nuclearsandwich committed:
Fix typo in repository name.

Co-authored-by: Jose Luis Rivero <jrivero@osrfoundation.org>

nuclearsandwich committed:
Trim trailing newlines

nuclearsandwich committed:
Add upstream_config for reprepro-updater.ini.

Since the reprepro-updater.ini file references the reprepro_config
directory I've moved the reprepro-updater.ini resource below the
reprepro_config resources and fixed the missing reference to it as teh
upstream_config field in reprepro-updater.ini

j-rivero reviewed (approved):
Minor discussion about permissions, does not seem important enough to block the PR. Looks good.

nuclearsandwich committed:
Authorize the publish over ssh key on the repo host.

nuclearsandwich committed:
Make gpg config readable only by owner.

Co-authored-by: Jose Luis Rivero <jrivero@osrfoundation.org>

nuclearsandwich committed:
Explicit sync action for clarity.

Sync is the default action for git resources but to emphasize that it is
the behavior we want there's no harm in declaring it explicitly.

nuclearsandwich committed:
Restrict ssh and gpg directories to owner-access only.

Co-authored-by: Jose Luis Rivero <jrivero@osrfoundation.org>
